plutus-core/executables/plutus/AnyProgram/With.hs

```diff
@@ -52,34 +65,71 @@ withA s r = case s of
+        )
+    => SLang s -> (constr (FromLang s) => r) -> r
+withLangGeneral s r = case s of
+    SPir sname sann ->
+          withNT @constrTyName sname
+        $ withN @constrName sname
+        $ withA @constrAnn sann r
+    SPlc sname sann ->
+          withNT @constrTyName sname
+        $ withN @constrName sname
+        $ withA @constrAnn sann r
+    SUplc sname sann ->
+          withN @(ComposeC constrBinder UPLC.Binder) sname
+        $ withN @constrName sname
+        $ withA @constrAnn sann r
+    SData -> error "not implemented yet"
 
-withShowL :: forall s r. SLang s ->  (Show (FromLang s) => r) -> r
-withShowL s r = case s of
-    SPir sname sann -> withN @Show sname $ withNT @Show sname $ withA @Show sann r
-    SPlc sname sann -> withN @Show sname $ withNT @Show sname $ withA @Show sann r
-    SUplc sname sann -> withN @Show sname $ withN @(ComposeC Show UPLC.Binder) sname $
-                       withA @Show sann r
-    SData -> r
+withLang
+    :: forall constr s r.
+        ( constr (FromNameTy 'Name)
+        , constr (FromNameTy 'DeBruijn)
+        , constr (FromNameTy 'NamedDeBruijn)
+        , constr (UPLC.Binder UPLC.Name)
+        , constr (UPLC.Binder UPLC.DeBruijn)
+        , constr (UPLC.Binder UPLC.NamedDeBruijn)
+        , constr (FromName 'Name)
+        , constr (FromName 'DeBruijn)
+        , constr (FromName 'NamedDeBruijn)
+        , constr (FromAnn 'Unit)
+        , constr (FromAnn 'TxSrcSpans)
+        , (forall tyname name ann. (constr tyname, constr name, constr ann) =>
+            constr (PIR.Program tyname name UPLC.DefaultUni UPLC.DefaultFun ann))
+        , (forall tyname name ann. (constr tyname, constr name, constr ann) =>
+            constr (PLC.Program tyname name UPLC.DefaultUni UPLC.DefaultFun ann))
+        , (forall name ann. (constr (UPLC.Binder name), constr name, constr ann) =>
+            constr (UPLC.UnrestrictedProgram name UPLC.DefaultUni UPLC.DefaultFun ann))
+        )
+    => SLang s -> (constr (FromLang s) => r) -> r
+withLang = withLangGeneral @constr @constr @constr @constr @constr
 
-withPrettyPlcL :: forall s r. SLang s -> (PrettyBy PrettyConfigPlc (FromLang s) => r) -> r
-withPrettyPlcL s r = case s of
-    SPir sname sann -> withN @PrettyClassic sname $ withN @PrettyReadable sname $
-                      withNT @PrettyClassic sname $ withNT @PrettyReadable sname $
-                      withA @Pretty sann r
-    SPlc sname sann -> withN @PrettyClassic sname $ withN @PrettyReadable sname $
-                      withNT @PrettyClassic sname $ withNT @PrettyReadable sname $
-                      withA @Pretty sann r
-    SUplc sname sann -> withN @PrettyClassic sname $ withN @PrettyReadable sname $
-                       withA @Pretty sann r
-    SData -> r
+withPrettyPlcL :: forall s r. SLang s -> (PrettyPlc (FromLang s) => r) -> r
+withPrettyPlcL = withLangGeneral
+    @(PrettyClassic `AndC` PrettyReadable)
```

Comment:
@effectfully : Just out of curiosity: is `AndC` strictly necessary here?

Could it alternatively be done by decomposing to two `withLangGeneral` calls? Something like:

```
withPrettyPlcL = withLangGeneral  @PrettyClassic ....   $ withLangGeneral @PrettyReadable ...
```
?

Reply:
> is `AndC` strictly necessary here?

I believe so, you have to satisfy a qualified constraint `(a, b, c) => d` and you can't do that unless `a`, `b` and `c` are indeed as expected by `d` and so you have to provide the right `a`, `b` and `c` which forces you to use `AndC`. In particular, `PrettyPlc` can only hold if both `PrettyClassic` and `PrettyReadable` hold, so you can't just split the constraint for individual parts. I think it might be possible in theory to make it so that `PrettyPlc` is implied by some product of constraints and then satisfy those individually, but that's pointless, writing `AndC` is simply easier.